katdal/dataset.py

```diff
@@ -224,9 +233,15 @@ def subrange(self, first, last):
         if not (0 <= first < last <= self.num_chans):
             raise IndexError('channel indices out of range')
         channel_shift = (first + last) // 2 - self.num_chans // 2
+        num_chans = last - first
+        # We use self.bandwidth and self.num_chans to avoid rounding errors
+        # that might accumulate if channel_width is inexact.
+        centre_freq = self.centre_freq \
+            + channel_shift * self.bandwidth * self.sideband / self.num_chans,
         return SpectralWindow(
-            self.centre_freq + channel_shift * self.channel_width * self.sideband,
-            self.channel_width, last - first, self.product, self.sideband, self.band)
+            centre_freq, self.channel_width, num_chans,
+            self.product, self.sideband, self.band,
+            bandwidth=self.bandwidth * num_chans / self.num_chans)
```

Comment:
If `subrange` should ideally preserve `channel_width`, why override it with a `bandwidth` parameter?

Reply:
Because it should also ideally compute an accurate bandwidth. For example, if the origin bandwidth is 1000Hz and you select half the channels, you should end up with 500Hz. If there were 14 channels and you preserved channel_width, you'd end up with a bandwidth of 1000/14*7, which might have rounding errors. I've mostly been favouring bandwidth over channel_width because that's what is stored in telstate.

Reply:
And will preserving `channel_width` help preserve the channel frequencies? A good question whether the latter is more NB than bandwidth or not.

Reply:
Not necessarily, because floating point is't distributive i.e. a channel frequency previously calculated as `old_center + i * channel_width` is now calculated as `(old_center + shift * channel_width) + (i - shift) * channel_width`. I think the only way to ensure that subrange doesn't alter channel frequencies is to store a reference channel and frequency and keep those the same in the return value, rather than always making the centre the reference.

Reply:
I see... I guess that's the rationale behind e.g. the FITS WCS keywords CRVALx and CRPIXx. More design choices to ponder...